compiler/rustc_data_structures/src/thin_vec.rs

```diff
@@ -46,6 +56,53 @@ impl<T> ::std::ops::DerefMut for ThinVec<T> {
     }
 }
 
+impl<T> FromIterator<T> for ThinVec<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut iter = iter.into_iter();
+
+        match iter.next() {
+            None => ThinVec(None),
+            Some(first) => {
+                // + 1 for the first element
+                let mut vec = Vec::with_capacity(iter.size_hint().0 + 1);
+
+                vec.push(first);
+                vec.extend(iter);
+
+                ThinVec(Some(Box::new(vec)))
+            }
+        }
```

Comment:
`Vec::from_iter` has a boatload of optimizations, so you might want to use `iter.into_iter().collect::<Vec<_>>()` instead and return `ThinVec(None)` if  `vec.is_empty()`. If the iterator is empty the `collect()` won't have allocated either.

Reply:
The one downside of that is that I think the `Vec` will still take up 16 extra bytes of stack space in this method, but maybe the other benefits are worth it.

Reply:
The stack space is fine - the reason for ThinVec is when you're storing lots of different vecs and the memory usage becomes significant. It can also affect the ABI of a hot path if the compiler has to spill things to the stack when it runs out of registers. But here the allocation takes far more time than the calling convention.

Reply:
Done.